SYCL/ESIMD/regression/globals.cpp

```diff
@@ -0,0 +1,162 @@
+    if (Id[2] + 1 < R[2]) {
+      Id[2]++;
+      return;
+    }
+    Id[2] = 0;
+  }
+  if constexpr (Dim >= 2) {
+    if (Id[1] + 1 < R[1]) {
+      Id[1]++;
+      return;
+    }
+    Id[1] = 0;
+  }
+  if (Id[0] + 1 < R[0])
+    Id[0]++;
+}
+
+template <int Dim> int testXD(queue &Q, nd_range<Dim> NDR) {
+  device Dev = Q.get_device();
+  context Context = Q.get_context();
+
+  std::cout << "Start test case for range: {" << NDR.get_global_range() << ", "
+            << NDR.get_local_range() << "}" << std::endl;
+
+  size_t GlobalRange = NDR.get_global_range().size();
+  size_t LocalRange = NDR.get_local_range().size();
+
+  id<Dim> *GlobalId = sycl::malloc_shared<id<Dim>>(GlobalRange, Dev, Context);
+  range<Dim> *GlobalSize =
+      sycl::malloc_shared<range<Dim>>(GlobalRange, Dev, Context);
+
+  id<Dim> *LocalId = sycl::malloc_shared<id<Dim>>(GlobalRange, Dev, Context);
+  range<Dim> *LocalSize =
+      sycl::malloc_shared<range<Dim>>(GlobalRange, Dev, Context);
+
+  uint32_t *SubGroupLocalId =
+      sycl::malloc_shared<uint32_t>(GlobalRange, Dev, Context);
+  uint32_t *SubGroupSize =
+      sycl::malloc_shared<uint32_t>(GlobalRange, Dev, Context);
+  uint32_t *SubGroupMaxSize =
+      sycl::malloc_shared<uint32_t>(GlobalRange, Dev, Context);
+
+  id<Dim> *GroupId = sycl::malloc_shared<id<Dim>>(GlobalRange, Dev, Context);
+  range<Dim> *NumGroups =
+      sycl::malloc_shared<range<Dim>>(GlobalRange, Dev, Context);
+
+  Q.submit([&](sycl::handler &CGH) {
+     namespace seoe = sycl::ext::oneapi::experimental;
+     CGH.parallel_for(NDR, [=](sycl::nd_item<Dim> NdId) SYCL_ESIMD_KERNEL {
+       size_t I = NdId.get_global_linear_id();
+       id<Dim> Id = NdId.get_global_id();
+
+       GlobalId[I] = Id;
+       GlobalSize[I] = NdId.get_global_range();
+
+       LocalId[I] = NdId.get_local_id();
+       LocalSize[I] = NdId.get_local_range();
+
+       SubGroupLocalId[I] = seoe::this_sub_group().get_local_id();
+       SubGroupSize[I] = seoe::this_sub_group().get_local_range()[0];
+       SubGroupMaxSize[I] = seoe::this_sub_group().get_max_local_range()[0];
+
+       GroupId[I] = NdId.get_group().get_id();
+       NumGroups[I] = NdId.get_group_range();
+     });
+   }).wait();
+
+  id<Dim> Id = zeroId<Dim>();
+  for (size_t I = 0; I < NDR.get_global_range().size(); I++) {
+    if (!check("GlobalId", Id, GlobalId[I], I) ||
+        !check("GlobalSize", NDR.get_global_range(), GlobalSize[I], I) ||
+        !check("LocalId", Id % NDR.get_local_range(), LocalId[I], I) ||
+        !check("LocalSize", NDR.get_local_range(), LocalSize[I], I) ||
+        !check("SubGroupLocalId", 0, SubGroupLocalId[I], I) ||
+        !check("SubGroupSize", 1, SubGroupSize[I], I) ||
+        !check("SubGroupMaxSize", 1, SubGroupMaxSize[I], I) ||
+        !check("GroupId", Id / NDR.get_local_range(), GroupId[I], I) ||
+        !check("NumGroups", NDR.get_group_range(), NumGroups[I], I))
+      return 1;
+    advanceId(Id, NDR.get_global_range());
+  }
+  return 0;
```

Comment:
It would be good to cleanup - sycl::free the allocated memory

Reply:
Thank you. Fixed in https://github.com/intel/llvm-test-suite/pull/503/commits/ba6e25fe14538093bc0b03fe53e63022250d4584